homeassistant/components/velbus/__init__.py

```diff
@@ -1,81 +1,44 @@
 """Support for Velbus devices."""
 import logging
 
-import velbus
+from velbusaio.controller import Velbus
 import voluptuous as vol
 
-from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
-from homeassistant.const import CONF_ADDRESS, CONF_NAME, CONF_PORT
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, CONF_PORT
 from homeassistant.core import HomeAssistant
-from homeassistant.exceptions import ConfigEntryNotReady
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.entity import Entity
 
 from .const import CONF_MEMO_TEXT, DOMAIN, SERVICE_SET_MEMO_TEXT
 
 _LOGGER = logging.getLogger(__name__)
 
-VELBUS_MESSAGE = "velbus.message"
-
 CONFIG_SCHEMA = vol.Schema(
     {DOMAIN: vol.Schema({vol.Required(CONF_PORT): cv.string})}, extra=vol.ALLOW_EXTRA
 )
 
 PLATFORMS = ["switch", "sensor", "binary_sensor", "cover", "climate", "light"]
 
 
-async def async_setup(hass, config):
-    """Set up the Velbus platform."""
-    # Import from the configuration file if needed
-    if DOMAIN not in config:
-        return True
-    port = config[DOMAIN].get(CONF_PORT)
-    data = {}
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Establish connection with velbus."""
+    hass.data.setdefault(DOMAIN, {})
 
-    if port:
-        data = {CONF_PORT: port, CONF_NAME: "Velbus import"}
-    hass.async_create_task(
-        hass.config_entries.flow.async_init(
-            DOMAIN, context={"source": SOURCE_IMPORT}, data=data
-        )
-    )
+    controller = Velbus(entry.data[CONF_PORT])
+    hass.data[DOMAIN][entry.entry_id] = controller
+    await controller.connect()
 
-    return True
+    for platform in PLATFORMS:
+        hass.add_job(hass.config_entries.async_forward_entry_setup(entry, platform))
 
+    async def scan(self, service=None):
+        await controller.scan()
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-    """Establish connection with velbus."""
-    hass.data.setdefault(DOMAIN, {})
+    hass.services.async_register(DOMAIN, "scan", scan, schema=vol.Schema({}))
```

Comment:
If the service has no data parameters we don't need to register a validation schema.
```suggestion
    hass.services.async_register(DOMAIN, "scan", scan)
```

Reply:
Note that it doesn't work to set up more than one config entry for this integration currently and call the services. Services are unique per service names, so we're currently overwriting the services each time a config entry is setup. We need to keep track of which controller to call when calling a service. We should probably fix this in a separate PR.

Reply:
do you have an example, or is this described somewehre how to do this?

Reply:
First we need to decide how to identify the controller from the service call data parameters. Maybe via a port parameter or if we create a Home Assistant device for the controller we could accept a device id via the device selector.

https://developers.home-assistant.io/docs/dev_101_services#service-descriptions

Reply:
this doc talks about a service linked to an entity and a platform.
But in this case i basicaly want to bind a service to the config_entry

Reply:
Yes. The service descriptions may work for this use case too. See eg the system_bridge integration. The controller in our case would correspond to the bridge in that case.

https://github.com/home-assistant/core/blob/49a69d5ba093c3ad6d95667632e866b44f9f8adc/homeassistant/components/system_bridge/services.yaml#L1-L11

https://github.com/home-assistant/core/blob/49a69d5ba093c3ad6d95667632e866b44f9f8adc/homeassistant/components/system_bridge/__init__.py#L114-L132